src/include/daos_srv/dtx_srv.h

```diff
@@ -96,23 +104,40 @@ enum dtx_status {
 	DTX_ST_COMMITTED	= 3,
 };
 
-/**
- * Some actions to be done for DTX control.
- */
-enum dtx_actions {
-	/** Need to aggregate some old DTXs. */
-	DTX_ACT_AGGREGATE	= 1,
-	/** Need to commit some old DTXs asychronously. */
-	DTX_ACT_COMMIT_ASYNC	= 2,
-	/** Commit current DTX sychronously. */
-	DTX_ACT_COMMIT_SYNC	= 3,
-};
-
-int dtx_commit(uuid_t po_uuid, uuid_t co_uuid,
-	       struct daos_tx_entry *dtes, int count, uint32_t version);
-int dtx_abort(uuid_t po_uuid, uuid_t co_uuid,
-	      struct daos_tx_entry *dtes, int count, uint32_t version);
 int dtx_resync(daos_handle_t po_hdl, uuid_t po_uuid, uuid_t co_uuid,
 	       uint32_t ver, bool block);
 
+int dtx_begin(struct dtx_id *dti, daos_unit_oid_t *oid, daos_handle_t coh,
+	      daos_epoch_t epoch, uint64_t dkey_hash,
+	      struct dtx_conflict_entry *conflict, struct dtx_id *dti_cos,
+	      int dti_cos_count, uint32_t pm_ver, uint32_t intent, bool leader,
+	      struct dtx_handle **dth);
+
+int dtx_end(struct dtx_handle *dth, struct ds_cont_hdl *cont_hdl,
+	    struct ds_cont *cont, int result);
+
+int dtx_conflict(daos_handle_t coh, daos_unit_oid_t *old, uuid_t po_uuid,
+		 uuid_t co_uuid, struct dtx_conflict_entry *dces, int count,
+		 uint32_t version);
+
+int dtx_batched_commit_register(struct ds_cont_hdl *hdl);
+
+void dtx_batched_commit_degister(struct ds_cont_hdl *hdl);
+
+int dtx_handle_resend(daos_handle_t coh, daos_unit_oid_t *oid,
+		      struct dtx_id *dti, uint64_t dkey_hash,
+		      uint64_t start_time, bool punch);
+
+/* XXX: It should be exported by CART. This is just temporary solution. */
+#ifndef CRT_HLC_MASK
+# define CRT_HLC_MASK	0xFFFFULL
```

Comment:
what is it for?  To just mask off the logical bits?   I don't think this is necessary.   The logical bits will amount to less than 1 second of age when divided by NSEC_PER_SEC

Reply:
That depends on how the HLC timestamp is generated, we cannot assume how the lower bits will be used.

Reply:
Why not?  We can assume that the clock is a 64-bit value with high order bits representing physical time and low order bits representing logical time starting at the given physical time.   How many lower order bits we have doesn't really matter at the granularities we are talking about here.

And any other bit ordering or usage would render HLC useless for comparison.   For example, if you put the logical bits at the high end, then you lose the ordering in simple comparisons (unless you have a callback to interpret the comparison between two time stamps).

So I think we can safely assume that the error is no more than 1 second and since your timeouts are in the 1000's of seconds, I don't think that's much of a problem.  Plus, we control CaRT

Reply:
OK, I will drop the unnecessary logic.